db/dbformat.h

```diff
@@ -504,6 +516,39 @@ class IterKey {
     key_size_ = total_size;
   }
 
+  // A version of `TrimAppend` assuming the last bytes of length
+  // `padded_user_key_len` in the user key part of `key_` is not counted towards
+  // shared bytes. And the decoded key needed a min timestamp of length
+  // `padded_user_key_len` pad to the user key.
+  void TrimAppendWithTimestamp(const size_t shared_len,
+                               const char* non_shared_data,
+                               const size_t non_shared_len,
+                               const size_t padded_user_key_len) {
+    std::string key_with_ts;
+    if (IsUserKey()) {
+      TrimAppend(shared_len, non_shared_data, non_shared_len);
+      AppendKeyWithMinTimestamp(&key_with_ts, GetKey(), padded_user_key_len);
+    } else {
+      std::string internal_key_without_ts;
+      // Invaraint: shared_user_key_len + shared_internal_bytes_len = shared_len
+      size_t sharable_user_key_len =
+          key_size_ - kNumInternalBytes - padded_user_key_len;
+      size_t shared_user_key_len = std::min(shared_len, sharable_user_key_len);
+      internal_key_without_ts.append(key_, shared_user_key_len);
+      // Some shared bytes come from the last few internal bytes.
+      if (shared_len > sharable_user_key_len) {
+        size_t shared_internal_bytes_len = shared_len - sharable_user_key_len;
+        size_t user_key_len = key_size_ - kNumInternalBytes;
+        internal_key_without_ts.append(key_ + user_key_len,
+                                       shared_internal_bytes_len);
+      }
+      internal_key_without_ts.append(non_shared_data, non_shared_len);
+      PadInternalKeyWithMinTimestamp(&key_with_ts, internal_key_without_ts,
+                                     padded_user_key_len);
```

Comment:
I tried to save some copying but didn't have any good ideas. Thanks a lot for the SliceParts idea, I have updated this logic to piece together the final key using SliceParts coming from its original sources. For both user key and internal key cases, there is 1) construct the key in a temp string buffer with different slice parts, 2) copy the content in temp string buffer via `SetKey`. 

This code becomes a bit chunky, let me know if it looks OK or need more polishing.

Reply:
Sorry, there's quite a bit of nuance here so I might be missing something but I'm wondering if something simpler like the following would work for internal keys (using the length variables you introduced):

```
key_parts_with_ts.emplace_back(key_, shared_user_key_len);
key_parts_with_ts.emplace_back(kTsMin);
if (shared_len > sharable_user_key_len) {
   key_parts_with_ts.emplace_back(key_ + user_key_len, shared_internal_bytes_len);
}
key_parts_with_ts.emplace_back(non_shared_data, non_shared_len);
```

If this works, `key_parts_with_ts` can be a `vector` (as opposed to a `deque`).

Reply:
Thank you for the suggestion! In this snippet, this invariant `shared_internal_bytes_len + non_shared_len == kNumInternalBytes` needs to always be true. 

This may not be true for all the cases.  1) when `shared_internal_bytes_len + non_shared_len < kNumInternalBytes`, the timestamp needs to be squeezed in between `Slice(key_, shared_user_key_len)`.  2) when `non_shared_len > kNumInternalBytes`, the min timestamp need to be squeezed in between `Slice(non_shared_data, non_shared_len)`. 3) there is also a case when the min timestamp need to be squeezed in between `Slice(key_ + user_key_len, shared_internal_bytes_len)`. 

I think all three cases potentially can happen, let me know if I miss something. Thanks for the `emplace_back` example in the snippet, I updated to use it instead, it helps visually declutter a little bit.

Reply:
Ugh. Right. It *is* more complicated than what I wrote :)

We should still be able to build the structure in one shot I think. Handwavy pseudocode assuming we have calculated how the shared and non-shared bytes are split up between the user key and the footer:

```
if (shared_user_key_bytes > 0) {
   key_parts_with_ts.emplace_back(/* ... */);
}
if (non_shared_user_key_bytes > 0) {
   key_parts_with_ts.emplace_back(/* ... */);
}
key_parts_with_ts.emplace_back(kTsMin);
if (shared_internal_bytes > 0) {
   key_parts_with_ts.emplace_back(/* ... */);
}
if (non_shared_internal_bytes > 0) {
   key_parts_with_ts.emplace_back(/* ... */);
}
```

What do you think?

Reply:
Thanks for the idea to get rid of the deque. We may be able to do this using hard coded if else statements like this, but I think it's a bit more complicated than this.

What I mean by "squeeze the min timestamp in between a Slice" is that the slice needs to be broken into two pieces, with the piece on the left to be added before the min timestamp and the piece on the right to be added after the min timestamp. Which Slice needs to be split and how to split it will depend on these length variable conditions. Let me know if I should do this instead of using the deque.

Reply:
This is a frequently hit piece of code, the performance of this piece is pretty important. Let me try out this idea to remove the deque.

Reply:
I made some changes and removed the deque. Let me know if this looks OK.

Reply:
> The final key comes from these three parts, plus the min timestamp to be padded: Slice 1 (key_, shared_user_key_len) Slice 2 (key_+user_key_len, shared_internal_bytes_len) Slice 3 (non_shared_data, non_shared_len)
> 
> In the recommended structure, the statement `key_parts_with_ts.emplace_back(kTsMin);` is in between the statements that added the user key parts, and the statements that added the footer. Depending on the lengths of these three parts: a) timestamp could be inserted in between the first Slice. In that case, the left part of Slice 1 is user key, while the right part of Slice 1 is footer, Slice 2 and Slice 3 are both footer. b) timestamp could be inserted in between the second Slice. Similarly, Slice 1 is user key, left part of Slice 2 is user key, right part of Slice 2 and Slice 3 are footer. c) timestamp could be inserted in between the third Slice, a different distribution between user key bytes and internal bytes.

Sorry about all these questions but I'm getting a bit confused now... I understand splitting slice 3 (the non-shared part) but why would we have to split slice 1 or 2? Slice 1 points to some bytes from the user-provided key, which should thus be strictly before the timestamp, while slice 2 points to bytes from the footer, which should be strictly after the timestamp, no?

Reply:
... and finding out where to split slice 3 is equivalent to computing `non_shared_user_key_bytes` and `non_shared_internal_bytes` in the above snippet.

Reply:
This is the reverse of the encoding process. During encoding, internal key's timestamp is stripped and then delta encoded. So during decoding, we first delta decode and then pad the timestamp.

For example, assuming `key_` currently holds this: "ab\0\0\0\0\0\0\0\012345678", "ab" is user key without timestamp, "12345678" are the footer, in between is the artificially padded timestamp.
If next key has something like: shared_len = 5 bytes, non_shared_len = 4, non_shared_data = "cdef".

The next key would be squeezing "\0\0\0\0\0\0\0\0" into "ab123cdef" leaving the last 8 bytes be the footer, the final key is: "a\0\0\0\0\0\0\0\0b123cdef".  In this case, Slice 1 is split to add the timestamp.

Reply:
> For example, assuming `key_` currently holds this: "ab\0\0\0\0\0\0\0\012345678", "ab" is user key without timestamp, "12345678" are the footer, in between is the artificially padded timestamp. If next key has something like: shared_len = 5 bytes, non_shared_len = 4, non_shared_data = "cdef".
> 
> The next key would be squeezing "\0\0\0\0\0\0\0\0" into "ab123cdef" leaving the last 8 bytes be the footer, the final key is: "a\0\0\0\0\0\0\0\0b123cdef". In this case, Slice 1 is split to add the timestamp.

Thanks a lot for the example! Working through it, the current `key_size_` would be 18 bytes, `user_key_len` would be 10 bytes, and `sharable_user_key_len` and `shared_user_key_len` would be 2 bytes. This does not seem correct, since the part that's shared between the two user keys is only 1 byte `'a'`; `b` is "moved" to the footer, which is technically the reason slice 1 has to be split.

I think we should be able to figure out the right splits up front working backwards through the internal key. Since `shared_len + non_shared_len == 9`, and the footer is 8 bytes, we should be able to cap `shared_user_key_len` at 1 byte (since that's the max length of the user key portion of the new key).

Reply:
This does get tricky though, with the same shared bytes forming a part of the user key in one key and the internal key in another... so the complexity might be unavoidable.

Reply:
Thank you for the suggestion of "figuring out the split up front working backwards through the internal key".  

I want to first explain the variable naming: `sharable_user_key_len` and `shared_user_key_len`, in both names, `user_key` refers to the fact that these bytes are part of the user key in the existing `key_`, not that it will be part of the user key for the new key. I didn't clearly mention this, it caused some confusion when we are discussing the three Slices.

About the idea of "figure out the right split", this is kind of what the abstracted function `MaybeAddKeyPartsWithTimestamp` is doing while handling the "complexity of the same shared bytes forming a part of the user key in one key and the internal key in another". It will determine which Slice to split depending on these conditions:
1) `shared_internal_bytes_len + non_shared_len < kNumInternalBytes` => the first Slice should be split, if not, continue to 2)
2) `non_shared_len < kNumInternalBytes` => the second Slice should be split, if not continue to 3)
3) `non_shared_len >= kNumInternalBytes` => the third Slice should be split

As long as `shared_len + non_shared_len >= kNumInternalBytes`, one of the three conditions has to be true, we assert this at the beginning, and assert `ts_added` at the end.

Reply:
> I want to first explain the variable naming: `sharable_user_key_len` and `shared_user_key_len`, in both names, `user_key` refers to the fact that these bytes are part of the user key in the existing `key_`, not that it will be part of the user key for the new key. I didn't clearly mention this, it caused some confusion when we are discussing the three Slices.

Yeah, I just realized this myself; with shared bytes potentially moving between the user key part and the footer, concepts like "shared user bytes" and "shared internal bytes" become ambiguous. It might make sense to add a comment explaining the semantics.

Reply:
... otherwise, LGTM! Sorry about the churn!

Reply:
No worries! Thanks a lot for the idea of `SliceParts` to eliminate extra copying, that checks an item off of my worry list. And also appreciate the ideas to optimize the performance of this piece of code. Deque is definitely an overkill for something that is known to have a maximum of three items in it.